backend/src/type_checking.py

```diff
@@ -0,0 +1,208 @@
+from __future__ import annotations
+
+import ast
+import inspect
+from typing import Any, Callable, Dict, List, NewType, Set, Union, cast, get_args
+
+from custom_types import NodeType
+from nodes.properties.inputs.base_input import BaseInput
+from nodes.properties.outputs.base_output import BaseOutput
+
+_Ty = NewType("_Ty", object)
+
+
+class TypeMismatchError(Exception):
+    pass
+
+
+class TypeTransformer(ast.NodeTransformer):
+    def visit_BinOp(self, node: ast.BinOp):
+        if isinstance(node.op, ast.BitOr):
+            return ast.Subscript(
+                value=ast.Name(id="Union", ctx=ast.Load()),
+                slice=ast.Index(
+                    value=ast.Tuple(
+                        elts=[
+                            self.visit(node.left),
+                            self.visit(node.right),
+                        ],
+                        ctx=ast.Load(),
+                    ),
+                    ctx=ast.Load(),
+                ),
+                ctx=ast.Load(),
+            )
+        return super().visit_BinOp(node)
+
+
+def compile_type_string(s: str, filename: str = "<string>"):
+    tree = ast.parse(s, filename, "eval")
+    new_tree = ast.fix_missing_locations(TypeTransformer().visit(tree))
+    return compile(new_tree, filename, "eval")
+
+
+def eval_type(t: str | _Ty, __globals: dict[str, Any]):
+    if not isinstance(t, str):
+        return t
+
+    # `compile_type_string` adds `Union`, so we need it in scope
+    local_scope = {
+        "Union": Union,
+    }
+
+    try:
+        # pylint: disable=eval-used
+        return _Ty(eval(compile_type_string(t), __globals, local_scope))
+    except Exception as e:
+        raise ValueError(f"Unable to evaluate type '{t}': {e}") from e
+
+
+def union_types(types: List[_Ty]) -> _Ty:
+    assert len(types) > 0
+    t: Any = types[0]
+    for t2 in types[1:]:
+        t = Union[t, cast(Any, t2)]
+    return t
+
+
+def union_to_set(t: _Ty) -> Set[_Ty]:
+    if str(t).startswith("typing.Union["):
+        return set(get_args(t))
+    else:
+        return {t}
```

Comment:
The fix for what we talked about on Discord.

```suggestion
def union_to_set(t: _Ty) -> Set[_Ty]:
    s = str(t)
    if s.startswith("typing.Union["):
        return set(get_args(t))
    elif s.startswith("typing.Optional["):
        return {*union_to_set(get_args(t)[0]), _Ty(type(None))}
    else:
        return {t}
```